### ISSUE TITLE
推理速度

### ISSUE THREAD

**Issue by Wuhuddy** [THUDM/VisionReward#4]
你好，推理的时候输入768x1024的图片，大概要一分半钟出结果
另外会出现重复回答的问题

![Image](https://github.com/user-attachments/assets/1cdaeee8-757c-4807-9b13-c740dc92fea7)

请问如何解决？

**Comment by xujz18** [comment]
您好，这可能是由于当前推理代码作为开源版本存在一部分mismatch的地方，我已经请 @Uranusxer 检查一下问题，多谢

**Comment by Uranusxer** [comment]
您好，能否进一步提供详细代码和命令行信息？关于推理速度，部署模型本身需要1min左右的时间，单个问题则仅需1-2s，可以部署模型后批量推理；另外对同一个图片/视频做多次问答时，可以对图像/视频embedding在transformer层做kv cache加速，我们未来也计划做进一步更新。

**Comment by gsfsdv** [comment]
# -*- encoding: utf-8 -*-
import os, sys
import json
import argparse
import torch
from tqdm import tqdm  # Import tqdm for the progress bar
from sat.model.mixins import CachedAutoregressiveMixin
from sat.quantization.kernels import quantize
from sat.model import AutoModel
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
from utils.utils import chat, llama2_tokenizer, llama2_text_processor_inference, get_image_processor
from utils.utils import llama3_tokenizer
from utils.models import CogVLMModel
from utils.models import VisualLlamaEVA
from io import BytesIO
import pandas as pd
from PIL import Image
import numpy as np
from VisionReward_Image.t2v_metrics.vqascore import VQAScore

MASK_INDICES = [0, 1, 2]      # Indices of mask features in original list
MASK_FEATURE_MAP = {
    0: [22, 23, 24, 28, 29],      # 'body(mask)' masks related features 'body correct' & 'harmfulness'
    1: [25, 26],                  # 'face(mask)' masks related features 'face'
    2: [27],                      # 'hands(mask)' masks related features 'hands'
}

def cal_score(args,image_path,prompt,model,text_processor_infer,image_processor):
    with open(args.ques_file, 'r') as file:
        ques_data = [line.strip() for line in file]
    with open(args.weight_file, 'r') as file2:
        weight_data = json.load(file2)
    wegiht = weight_data['coef']
    intercept = weight_data['intercept']
    answer_list = []
    alignment_score = VQAScore(model='clip-flant5-xxl') # our recommended scoring model
    alignment = alignment_score(images=[image_path], texts=[prompt])[0][0].cpu().item() 
    for ques in tqdm(ques_data, f'scoring image:{image_path}'):
        try:
            response, _, _ = chat(
                image_path=image_path,
                image = None,
                model=model,
                text_processor=text_processor_infer,
                img_processor=image_processor,
                query=ques,
                max_length=args.max_length,
                top_p=args.top_p,
                temperature=args.temperature,
                top_k=args.top_k,
                invalid_slices=text_processor_infer.invalid_slices,
                args=args
            )
            answer_list.append(response)
        except Exception as e:
            answer_list.append(None)
            print(f"Error processing {ques}: {str(e)}")
    reward = [(1 if ans =='yes<|end_of_text|>' else -1 ) for ans in answer_list]
    # add mask
    for mask_index, feature_indices in MASK_FEATURE_MAP.items():
        for feature_index in feature_indices:
            reward[feature_index] *= (int)(reward[mask_index] > 0)
    reward_filtered = [v for i, v in enumerate(reward) if i not in MASK_INDICES]
    final_reward = [alignment] + reward_filtered
    score = np.dot(final_reward, wegiht) + intercept
    return score[0]

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--max_length", type=int, default=3328, help='max length of the total sequence')
    parser.add_argument("--top_p", type=float, default=0.4, help='top p for nucleus sampling')
    parser.add_argument("--top_k", type=int, default=1, help='top k for top k sampling')
    parser.add_argument("--temperature", type=float, default=0.8, help='temperature for sampling')
    parser.add_argument("--version", type=str, default="vqa", choices=['chat', 'vqa', 'chat_old', 'base'], help='version of language process')
    parser.add_argument("--quant", choices=[8, 4], type=int, default=None, help='quantization bits')
    parser.add_argument("--from_pretrained", type=str, default="/mnt/dolphinfs/ssd_pool/docker/user/hadoop-aipnlp/HAL/hongshibo/models/VisionReward", help='pretrained ckpt')  # You need to first download the model from https://huggingface.co/THUDM/VisionReward-Image and then refer to its README to extract the checkpoint.
    parser.add_argument("--tokenizer_path", type=str, default="/mnt/dolphinfs/ssd_pool/docker/user/hadoop-aipnlp/HAL/hongshibo/models/Llama-3-8B-Instruct", help='tokenizer path')
    parser.add_argument("--fp16", action="store_true", help="Use fp16 precision")
    parser.add_argument("--bf16", action="store_true", help="Use bf16 precision")
    parser.add_argument("--stream_chat", action="store_true")
    parser.add_argument("--ques_file", type=str, default="VisionReward_Image/VisionReward_image_qa_select.txt", help="Path to the meta question file")
    parser.add_argument("--weight_file", type=str, default="VisionRewardImage/weight_select.json", help="Path to the weight file")
    parser.add_argument('--question', type=str, help='Question to be answered', default='Is the image clear?')
    parser.add_argument('--score', help='Whether to output the score', default=False, action='store_true')
    args = parser.parse_args()

    # Initialize model
    model, model_args = VisualLlamaEVA.from_pretrained(
        args.from_pretrained,
        args=argparse.Namespace(
            deepspeed=None,
            local_rank=0,
            rank=0,
            world_size=1,
            model_parallel_size=1,
            mode='inference',
            skip_init=True,
            use_gpu_initialization=not args.quant,
            device='cpu' if args.quant else 'cuda',
            **vars(args)
        )
    )
    model = model.eval()
    if args.quant:
        quantize(model, args.quant)
        if torch.cuda.is_available():
            model = model.cuda()
    model.add_mixin('auto-regressive', CachedAutoregressiveMixin())
    tokenizer = llama3_tokenizer(args.tokenizer_path, signal_type=args.version)
    image_processor = get_image_processor(model_args.eva_args["image_size"][0])
    text_processor_infer = llama2_text_processor_inference(tokenizer, args.max_length, model.image_length)
    
    # Set input
    image_path1 = "asset/test/test1.jpg"
    image_path2 = "asset/test/test2.jpg"
    prompt = "A child shouting on a chair with several plush toys."
    
    with torch.no_grad():
        if args.score:
            score = cal_score(args,image_path1,prompt,model,text_processor_infer,image_processor)
            print(f"score: {score}")
        else:
            ques = args.question
            response, _, _ = chat(
                image_path=image_path1,
                image = None,
                model=model,
                text_processor=text_processor_infer,
                img_processor=image_processor,
                query=ques,
                max_length=args.max_length,
                top_p=args.top_p,
                temperature=args.temperature,
                top_k=args.top_k,
                invalid_slices=text_processor_infer.invalid_slices,
                args=args
            )
            print(f"response:{response}")
               

if __name__ == "__main__":
    main()

使用上面的代码遇到同样的问题

**Comment by yuanygong** [comment]
请问有新的解决办法么，同样出现了这个问题

**Comment by Uranusxer** [comment]
抱歉回复较晚，我用@Wuhuddy 的代码仅修改tokenizer和model路径后运行输出正常。看上去问题是输出<|end_of_text|>后模型未停止输出，可能是环境（可以重新install下sat库）和模型ckpt（确保模型拼接无误）的问题。如果仍无法解决，只要确保模型正确，那模型的第一个token输出应该也是正确的，可以在sat.generation.autoregressive_sampling中新增filling_sequence_token函数如下：
```
def filling_sequence_token(
        model,
        seq,
        batch_size,
        strategy=BaseStrategy(),
        max_memory_length=100000,
        log_attention_weights=None,
        get_masks_and_position_ids=get_masks_and_position_ids_default,
        mems=None,
        max_new_tokens=1,  # 限制生成的最大 Token 数 （新添加的参数）
        **kw_args
        ):
    '''
        seq: [2, 3, 5, ..., -1(to be generated), -1, ...]
        mems: [num_layers, batch_size, len_mems(index), mem_hidden_size]
            cache, should be first mems.shape[1] parts of context_tokens.
            mems are the first-level citizens here, but we don't assume what is memorized.
            input mems are used when multi-phase generation.
    '''
    assert len(seq.shape) == 1
    if hasattr(strategy, 'num_beams') and batch_size < strategy.num_beams:
        batch_size = strategy.num_beams
        print_rank0(f'Adjust batch_size to {batch_size} due to num_beams. Mute this warning by setting batch_size == num_beams.', level='DEBUG')

    # Building the initial tokens, attention_mask, and position_ids
    context_length = 0
    while seq[context_length] >= 0:
        context_length += 1  # [0, context_length-1] are given
    assert context_length > 0
    tokens, attention_mask, position_ids = get_masks_and_position_ids(seq)
    tokens = tokens[..., :context_length]
    if attention_mask.dtype != torch.bool:
        attention_mask = attention_mask.type_as(next(model.parameters()))  # if fp16
    # Initialize generation
    counter = context_length - 1  # Last fixed index is ``counter''
    index = 0 if mems is None else mems.shape[2]  # Next forward starting index, also the length of cache.
    mems_cross = None
    new_tokens_generated = 0  # Track the number of new tokens generated

    # Step-by-step generation
    while counter < len(seq) - 1:
        # Now, we want to generate seq[counter + 1],
        # token[:, index: counter+1] needs forwarding.

        if seq[counter + 1] >= 0:  # Provided
            tokens = torch.cat(
                (
                    tokens,
                    seq[counter + 1: counter + 2].expand(tokens.shape[0], 1)
                ), dim=1
            )
            counter += 1
            continue

        # Forward
        if log_attention_weights is not None:
            log_attention_weights_part = log_attention_weights[..., index: counter + 1, :counter + 1]  # TODO memlen
        else:
            log_attention_weights_part = None

        if mems_cross is not None or 'encoder_outputs' not in kw_args:
            kw_args['encoder_outputs'] = mems_cross

        logits, *output_per_layers = model(
            input_ids=tokens[:, index:],
            position_ids=position_ids[..., index: counter + 1],
            attention_mask=attention_mask[..., index: counter + 1, :counter + 1],  # TODO memlen
            mems=mems,
            log_attention_weights=log_attention_weights_part,
            **kw_args
        )
        if len(output_per_layers) > 0 and 'mem_cross' in output_per_layers[0]:
            mems_cross = [mem['mem_cross'] for mem in output_per_layers]
        mem_kv = [o['mem_kv'] for o in output_per_layers]
        mems = update_mems(mem_kv, mems, max_memory_length=max_memory_length)
        counter += 1
        index = counter

        # Sampling
        logits = logits[:, -1].expand(batch_size, -1)  # [batch size, vocab size]
        tokens = tokens.expand(batch_size, -1)
        tokens, mems = strategy.forward(logits, tokens, mems)
        
        # After sampling a single token
        new_tokens_generated += 1
        if new_tokens_generated >= max_new_tokens:  # Check if the maximum number of new tokens is reached
            break  # Stop as soon as the first new token is generated

        if strategy.is_done:
            break

    return strategy.finalize(tokens, mems)
```
然后在VisionReward/utils/utils/chat.py的chat函数第158行调用filling_sequence_token而不是filling_sequence，让模型每次只输出一个token即可解决。

**Comment by GongXinyuu** [comment]
> 抱歉回复较晚，我用[@Wuhuddy](https://github.com/Wuhuddy) 的代码仅修改tokenizer和model路径后运行输出正常。看上去问题是输出<|end_of_text|>后模型未停止输出，可能是环境（可以重新install下sat库）和模型ckpt（确保模型拼接无误）的问题。如果仍无法解决，只要确保模型正确，那模型的第一个token输出应该也是正确的，可以在sat.generation.autoregressive_sampling中新增filling_sequence_token函数如下：
> 
> ```
> def filling_sequence_token(
>         model,
>         seq,
>         batch_size,
>         strategy=BaseStrategy(),
>         max_memory_length=100000,
>         log_attention_weights=None,
>         get_masks_and_position_ids=get_masks_and_position_ids_default,
>         mems=None,
>         max_new_tokens=1,  # 限制生成的最大 Token 数 （新添加的参数）
>         **kw_args
>         ):
>     '''
>         seq: [2, 3, 5, ..., -1(to be generated), -1, ...]
>         mems: [num_layers, batch_size, len_mems(index), mem_hidden_size]
>             cache, should be first mems.shape[1] parts of context_tokens.
>             mems are the first-level citizens here, but we don't assume what is memorized.
>             input mems are used when multi-phase generation.
>     '''
>     assert len(seq.shape) == 1
>     if hasattr(strategy, 'num_beams') and batch_size < strategy.num_beams:
>         batch_size = strategy.num_beams
>         print_rank0(f'Adjust batch_size to {batch_size} due to num_beams. Mute this warning by setting batch_size == num_beams.', level='DEBUG')
> 
>     # Building the initial tokens, attention_mask, and position_ids
>     context_length = 0
>     while seq[context_length] >= 0:
>         context_length += 1  # [0, context_length-1] are given
>     assert context_length > 0
>     tokens, attention_mask, position_ids = get_masks_and_position_ids(seq)
>     tokens = tokens[..., :context_length]
>     if attention_mask.dtype != torch.bool:
>         attention_mask = attention_mask.type_as(next(model.parameters()))  # if fp16
>     # Initialize generation
>     counter = context_length - 1  # Last fixed index is ``counter''
>     index = 0 if mems is None else mems.shape[2]  # Next forward starting index, also the length of cache.
>     mems_cross = None
>     new_tokens_generated = 0  # Track the number of new tokens generated
> 
>     # Step-by-step generation
>     while counter < len(seq) - 1:
>         # Now, we want to generate seq[counter + 1],
>         # token[:, index: counter+1] needs forwarding.
> 
>         if seq[counter + 1] >= 0:  # Provided
>             tokens = torch.cat(
>                 (
>                     tokens,
>                     seq[counter + 1: counter + 2].expand(tokens.shape[0], 1)
>                 ), dim=1
>             )
>             counter += 1
>             continue
> 
>         # Forward
>         if log_attention_weights is not None:
>             log_attention_weights_part = log_attention_weights[..., index: counter + 1, :counter + 1]  # TODO memlen
>         else:
>             log_attention_weights_part = None
> 
>         if mems_cross is not None or 'encoder_outputs' not in kw_args:
>             kw_args['encoder_outputs'] = mems_cross
> 
>         logits, *output_per_layers = model(
>             input_ids=tokens[:, index:],
>             position_ids=position_ids[..., index: counter + 1],
>             attention_mask=attention_mask[..., index: counter + 1, :counter + 1],  # TODO memlen
>             mems=mems,
>             log_attention_weights=log_attention_weights_part,
>             **kw_args
>         )
>         if len(output_per_layers) > 0 and 'mem_cross' in output_per_layers[0]:
>             mems_cross = [mem['mem_cross'] for mem in output_per_layers]
>         mem_kv = [o['mem_kv'] for o in output_per_layers]
>         mems = update_mems(mem_kv, mems, max_memory_length=max_memory_length)
>         counter += 1
>         index = counter
> 
>         # Sampling
>         logits = logits[:, -1].expand(batch_size, -1)  # [batch size, vocab size]
>         tokens = tokens.expand(batch_size, -1)
>         tokens, mems = strategy.forward(logits, tokens, mems)
>         
>         # After sampling a single token
>         new_tokens_generated += 1
>         if new_tokens_generated >= max_new_tokens:  # Check if the maximum number of new tokens is reached
>             break  # Stop as soon as the first new token is generated
> 
>         if strategy.is_done:
>             break
> 
>     return strategy.finalize(tokens, mems)
> ```
> 
> 然后在VisionReward/utils/utils/chat.py的chat函数第158行调用filling_sequence_token而不是filling_sequence，让模型每次只输出一个token即可解决。

这样修改后可以解决这个问题了，不过这一行应该需要修改一下：https://github.com/THUDM/VisionReward/blob/26c6508726e13263acda2926d558cc2183a6c95c/inference-image.py#L58
改为
```python
reward = [(1 if ans =='yes' else -1 ) for ans in answer_list]
```

**Comment by Uranusxer** [comment]
嗯嗯是的，这里转换分数的时候也需要根据模型输出稍微改一下

**Comment by liangbingzhao** [comment]
请问为什么我用inference-image.py 推理一张1024*1024的图片，要快半小时啊；用的bf16 ckpt
